core-common/src/main/java/org/mapstruct/BeforeMapping.java

```diff
@@ -0,0 +1,108 @@
+/**
+ *  Copyright 2012-2015 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks a method to be invoked at the beginning of a generated mapping method. The method can be implemented in an
+ * abstract mapper class or be declared in a type (class or interface) referenced in {@link Mapper#uses()} in order to
+ * be used in a mapping method.
+ * <p>
+ * Only methods with return type {@code void} may be annotated with this annotation.
+ * <p>
+ * If the method has parameters, the method invocation is only generated if all parameters can be <em>assigned</em> by
+ * the source or target parameters of the mapping method:
+ * <ul>
+ * <li>A parameter annotated with {@code @}{@link MappingTarget} is populated with the target instance of the mapping.
+ * </li>
+ * <li>A parameter annotated with {@code @}{@link TargetType} is populated with the target type of the mapping.</li>
+ * <li>Any other parameter is populated with a source parameter of the mapping, whereas each source parameter is used
+ * once at most.</li>
+ * </ul>
+ * If a <em>before-mapping</em> method does not contain a {@code @}{@link MappingTarget} parameter, it is invoked
+ * directly at the beginning of the applicable mapping method. If it contains a {@code @}{@link MappingTarget}
+ * parameter, the method is invoked after the target parameter has been initialized in the mapping method.
```

Comment:
Not necessarily - the `@BeforeMapping` method with the parameter list `(Source source, @MappingTarget target)` would be applicable to both `Target toTarget(Source s)` and `void intoTarget(Source s, @MappingTarget t)`. Because for the lifecycle method, it is only important to know - hey, there is going to be a bean mapping from Source to Target being done. `@MappingTarget` just denotes the instance to be returned by the mapping method.

To combine this with the previous comment, a use case could be some logging (for whatever reason), or if you'd want to perform a flush if you start mapping from an entity to a DTO, but not if you map from an entity into a JAXB type.